compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -1158,6 +1158,19 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                 );
             }
 
+            if path_segment.ident.name == sym::ok_or && is_diagnostic_item(sym::Option, next_ty) {
```

Comment:
This seems way too specific.

We already have logic to check for an "`_else`" version of the method. Why isn't that being triggered?

Also, this should actually be checking if the argument is something callable. Right now this will trigger for `.ok_or(non_closure_expression)`, no?

Also, I don't believe that the right place to put this expression is in `try_conversion_context`, since that means the only place it will apply is when we use the `?` operator.

Reply:
The current implementation does not trigger for a non-closure expression. Actually, we want this to apply only to the `?` operator, as this code compiles successfully:
```
fn main() -> Result<(), bool> {
    let _ = None::<bool>.ok_or(|| true);
    Ok(())
}
```

Reply:
Does your implementation work for this case that is not using `?`?

```rs
fn main() -> Result<(), bool> {
    if let Err(e) = None::<bool>.ok_or(|| true) {
        let b: bool = e;
    }
    Ok(())
}
```

Reply:
No